File: android-client/app/src/main/kotlin/com/lowremote/codec/FrameAssembler.kt
package com.lowremote.codec

import android.util.Log
import com.lowremote.model.Packet
import java.util.concurrent.ConcurrentHashMap

/**
 * Re-assembles UDP fragments (one Annex-B H.265 frame per frame_id) into
 * complete byte arrays ready for MediaCodec.
 *
 * At 120fps the inter-frame window is ~8ms; any fragment whose other pieces
 * don't arrive within [timeoutMs] is evicted — a late frame is worse than a
 * dropped one because it would cascade latency onto everything behind it.
 *
 * This class is safe to call from the UDP receive thread.
 */
class FrameAssembler(
    private val timeoutMs: Long = 50,
) {
    companion object { private const val TAG = "FrameAssembler" }

    private class FrameBuffer(val total: Int, val createdAt: Long) {
        val chunks = arrayOfNulls<ByteArray>(total)
        var received = 0
        var keyframe = false
        fun put(idx: Int, data: ByteArray) {
            if (idx < 0 || idx >= total) return
            if (chunks[idx] == null) {
                chunks[idx] = data
                received++
            }
        }
        val isComplete: Boolean get() = received == total
        fun assemble(): ByteArray {
            val totalLen = chunks.sumOf { it?.size ?: 0 }
            val out = ByteArray(totalLen)
            var pos = 0
            for (c in chunks) {
                if (c != null) {
                    System.arraycopy(c, 0, out, pos, c.size)
                    pos += c.size
                }
            }
            return out
        }
    }

    /** frame_id -> in-flight buffer */
    private val frames = ConcurrentHashMap<Int, FrameBuffer>()

    /** Highest frame_id we've already emitted, to drop late arrivals. */
    @Volatile private var lastEmittedId: Int = -1

    /** Callback invoked with a fully assembled frame (Annex-B). */
    var onFrameReady: ((bytes: ByteArray, isKeyframe: Boolean) -> Unit)? = null

    fun onPacket(parsed: Packet.Parsed, payload: ByteArray) {
        if (parsed.type != Packet.TYPE_VIDEO) return
        val frameId = parsed.frameId
        if (frameId <= lastEmittedId && !parsed.isKeyframe) {
            // Already emitted a later frame; this one's stale.
            return
        }

        if (parsed.pktTotal <= 0) return
        val buffer = frames.getOrPut(frameId) {
            FrameBuffer(parsed.pktTotal, System.currentTimeMillis())
        }
        buffer.put(parsed.pktIdx, payload)
        if (parsed.isKeyframe) buffer.keyframe = true

        if (buffer.isComplete) {
            frames.remove(frameId)
            val bytes = buffer.assemble()
            lastEmittedId = frameId
            onFrameReady?.invoke(bytes, buffer.keyframe)
            evictOlder(frameId)
        }
        evictExpired()
    }

    private fun evictOlder(currentId: Int) {
        // Drop any in-flight frame older than the one we just emitted.
        val iter = frames.entries.iterator()
        while (iter.hasNext()) {
            val e = iter.next()
            if (e.key < currentId) iter.remove()
        }
    }

    private fun evictExpired() {
        val now = System.currentTimeMillis()
        val iter = frames.entries.iterator()
        while (iter.hasNext()) {
            val e = iter.next()
            if (now - e.value.createdAt > timeoutMs) {
                iter.remove()
                Log.d(TAG, "dropped incomplete frame ${e.key} (${e.value.received}/${e.value.total})")
            }
        }
    }

    fun reset() {
        frames.clear()
        lastEmittedId = -1
    }
}

File: ios-client/LowRemote/Codec/FrameAssembler.swift
import Foundation

/// UDP 分片重组器，完整对齐 Android FrameAssembler.kt
///
/// 以 frameId 为 key 收集同一帧的所有分片，全部到齐后回调 onFrameReady。
/// 超时帧（默认 50ms，比 Android 8ms 宽松以适配 iOS 调度）自动清除。
final class FrameAssembler {

    // MARK: - Config
    private let timeoutMs: Int64 = 50
    private let maxPendingFrames = 32

    // MARK: - State
    private var frames: [UInt32: FrameBuffer] = [:]
    private let lock  = NSLock()
    private var lastEvictTime: Int64 = 0

    /// (完整帧字节, 是否关键帧) → 解码器
    var onFrameReady: ((Data, Bool) -> Void)?

    // MARK: - Public

    func onPacket(_ parsed: Packet.Parsed, _ payload: Data) {
        let now = currentMs()

        lock.lock()
        // 定期清过期帧（每 100ms 一次）
        if now - lastEvictTime > 100 {
            evictExpired(now: now)
            lastEvictTime = now
        }

        let fid = parsed.frameId
        if frames[fid] == nil {
            // 防止无限增长
            if frames.count >= maxPendingFrames { evictOldest() }
            frames[fid] = FrameBuffer(total: Int(parsed.pktTotal),
                                      isKeyframe: parsed.isKeyframe,
                                      createdAt: now)
        }

        let buf = frames[fid]!
        buf.put(index: Int(parsed.pktIdx), data: payload)

        if buf.isComplete {
            frames.removeValue(forKey: fid)
            lock.unlock()
            if let assembled = buf.assemble() {
                onFrameReady?(assembled, buf.isKeyframe)
            }
            return
        }
        lock.unlock()
    }

    func reset() {
        lock.lock()
        frames.removeAll()
        lock.unlock()
    }

    // MARK: - Private

    private func evictExpired(now: Int64) {
        frames = frames.filter { now - $0.value.createdAt <= timeoutMs }
    }

    private func evictOldest() {
        guard let oldest = frames.min(by: { $0.value.createdAt < $1.value.createdAt }) else { return }
        frames.removeValue(forKey: oldest.key)
    }

    private func currentMs() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }
}

// MARK: - FrameBuffer

private final class FrameBuffer {
    let total:      Int
    let isKeyframe: Bool
    let createdAt:  Int64

    private var slots:    [Data?]
    private var received: Int = 0

    init(total: Int, isKeyframe: Bool, createdAt: Int64) {
        self.total      = max(total, 1)
        self.isKeyframe = isKeyframe
        self.createdAt  = createdAt
        self.slots      = Array(repeating: nil, count: self.total)
    }

    var isComplete: Bool { received == total }

    func put(index: Int, data: Data) {
        guard index >= 0 && index < total else { return }
        if slots[index] == nil {
            slots[index] = data
            received += 1
        }
    }

    func assemble() -> Data? {
        var out = Data()
        for slot in slots {
            guard let s = slot else { return nil }
            out.append(s)
        }
        return out
    }
}
